examples/step-44/step-44.cc

```diff
@@ -2001,16 +1998,18 @@ namespace Step44
       {
         fe_values_ref.reinit(cell);
 
-        PointHistory<dim> *lqph =
-          reinterpret_cast<PointHistory<dim>*>(cell->user_pointer());
-
-        Assert(lqph >= &quadrature_point_history.front(), ExcInternalError());
-        Assert(lqph <= &quadrature_point_history.back(), ExcInternalError());
+        // In contrast to that which was previously called for,
+        // in this instance the quadrature point data is specifically
+        // non-modifiable since we will only be accessing data.
+        const std::vector<std::shared_ptr<const PointHistory<dim> > > lqph =
+            const_cast<const CellDataStorage<typename Triangulation<dim>::cell_iterator,
+                                             PointHistory<dim> >&>(quadrature_point_history).get_data(cell);
```

Comment:
I respectfully disagree. I don't see how this is any more complicated than the `reinterpret_cast` used in the previous version of the tutorial. Besides, its become more apparent that novice `C++` programmers use these tutorials to learn more about `C++` - whats wrong with imparting some knowledge? They can question it, play with it, and find out why this is necessary (or, perhaps when using another compiler, is not necessary). Besides, they'll have gotten up to at least `step-18` by this point, which means that they would have been exposed to a number of advanced techniques.

As to your second point, there's always going to be a clash here. 
- We could try to set `assemble_system_tangent_one_cell` as a constant function, but then `WorkStream::run` won't allow that because `*this` is non-const in `assemble_system_tangent`. We could then set all of these functions const and have `tangent_matrix` as a mutable data type, but that doesn't make sense because we expect assembly to change some member data.
- We could send in `tangent_matrix` into the assembly routine by non-const reference, and pass it though to `copy_local_to_global_K` and modify it. I think that this would work, but then one either needs to pass this object to `per_task_data` (this doesn't fit with what `per_task_data` is there for), or use a much more complex `WorkStream::run` call to customise the parameters passed to it. This is not a clean solution but would of course work.
- We could have a member function `get_const_qph() const` that returns a const reference to the `quadrature_point_history`. This would achieve the desired result but may elicit some head-scratching from the reader.
- We can do what we've done here, and ensure that something which is logically constant at this point in the code is interpreted as such, and the enforcement of this point is transparent. I think that we've taken the lesser of all evils here. But I do however note that in the previous version of the tutorial we did not ensure that data inside `lqph` would not be modified in the assembly calls.